In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os

In [2]:
path = r"D:\Data Immersion - Careerfoundry\Python Fundamentals for Data Analysts\Instacart Basket Analysis"

In [3]:
# Import orders_products_merged.pkl file
ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged.pkl'))

In [4]:
# Check the dataframe that was imported
ords_prods_merge.head()

Unnamed: 0_x  product_id                product_name  aisle_id  \
0             0           1  Chocolate Sandwich Cookies        61   
1             0           1  Chocolate Sandwich Cookies        61   
2             0           1  Chocolate Sandwich Cookies        61   
3             0           1  Chocolate Sandwich Cookies        61   
4             0           1  Chocolate Sandwich Cookies        61   

   department_id  prices  Unnamed: 0_y  order_id  user_id  order_number  \
0             19     5.8          1987   3139998      138            28   
1             19     5.8          1989   1977647      138            30   
2             19     5.8         11433    389851      709             2   
3             19     5.8         12198    652770      764             1   
4             19     5.8         12200   1813452      764             3   

   orders_day_of_the_week  order_hour_of_day  days_since_prior_order  \
0                       6                 11                     3.0   
1                       6                 17                    20.0   
2                       0                 21                     6.0   
3                       3                 13                     7.0   
4                       4                 17                     9.0   

   add_to_cart_order  reordered _merge    price_range_loc   busiest_day  \
0                  5          0   both  Mid Range product  Regular_busy   
1                  1          1   both  Mid Range product  Regular_busy   
2                 20          0   both  Mid Range product   Busiest day   
3                 10          0   both  Mid Range product  Regular_busy   
4                 11          1   both  Mid Range product    Least busy   

   busiest_days busiest_period_of_day  
0  Regular Busy           Most orders  
1  Regular Busy           Most orders  
2  Busiest days        Average orders  
3  Slowest days           Most orders  
4  Slowest days           Most orders

In [5]:
ords_prods_merge.shape

(32404859, 20)

# Task

## In this section you will find...

### Aggregating 'order_number' column grouped by 'department_id'

### Creating a flag for loyal customers and analyzing the results

### Creating a flag for spending of customers

### Creating a flag for frequency of orders by customers

### 2)

In [6]:
ords_prods_merge.groupby('department_id').agg({'order_number' : ['mean','min','max']})

order_number        
                      mean min max
department_id                     
1                15.457838   1  99
2                17.277920   1  99
3                17.170395   1  99
4                17.811403   1  99
5                15.215751   1  99
6                16.439806   1  99
7                17.225802   1  99
8                15.340650   1  99
9                15.895474   1  99
10               20.197148   1  99
11               16.170638   1  99
12               15.887671   1  99
13               16.583536   1  99
14               16.773669   1  99
15               16.165037   1  99
16               17.665606   1  99
17               15.694469   1  99
18               19.310397   1  99
19               17.177343   1  99
20               16.473447   1  99
21               22.902379   1  99

Overall, the results are not the same. Some departments had a lower average than from the subset, no more than 3 though. And other departments saw an increase in their mean. For example in department 10, it saw an increase some 18.6 to 20.2.

### 4)

In [4]:
# Create a order for the max number of orders from customers
ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)

In [5]:
# Create a flag column to determine a customers loyalty 
ords_prods_merge.loc[ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'
ords_prods_merge.loc[(ords_prods_merge['max_order'] <= 40) & (ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular Customer'
ords_prods_merge.loc[ords_prods_merge['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [9]:
# Check to see of column was created
ords_prods_merge[['max_order','loyalty_flag']].head()

max_order      loyalty_flag
0         32  Regular Customer
1         32  Regular Customer
2          5      New customer
3          3      New customer
4          3      New customer

In [10]:
ords_prods_merge['loyalty_flag'].value_counts(dropna = False)

Regular Customer    15876776
Loyal customer      10284093
New customer         6243990
Name: loyalty_flag, dtype: int64

In [14]:
ords_prods_merge.groupby('loyalty_flag').agg({'prices':['min','max','mean']})

prices                    
                    min      max       mean
loyalty_flag                               
Loyal customer      1.0  99999.0  10.386336
New customer        1.0  99999.0  13.294670
Regular Customer    1.0  99999.0  12.495717

### 5)

In [11]:
# Get basic stats on the prices of the various loyalty levels
ords_prods_merge.groupby('loyalty_flag').agg({'prices' : ['mean', 'min', 'max', 'median']})

prices                     
                       mean  min      max median
loyalty_flag                                    
Loyal customer    10.386336  1.0  99999.0    7.4
New customer      13.294670  1.0  99999.0    7.4
Regular Customer  12.495717  1.0  99999.0    7.4

As you can see there is a pricing difference between the different customers. It seems loyal customers tend to buy things at a lower price than other customers. This could be due to loyal customers looking for sales at the store and buying more items at a lower cost

### 6)

In [12]:
# Create flag column of customers spending
ords_prods_merge['avg_price'] = ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)

In [13]:
ords_prods_merge.loc[ords_prods_merge['avg_price'] < 10, 'spending_flag'] = 'Low spender'
ords_prods_merge.loc[ords_prods_merge['avg_price'] >= 10, 'spending_flag'] = 'High spender'

In [14]:
# Double to see if columns was created
ords_prods_merge.head()

Unnamed: 0_x  product_id                product_name  aisle_id  \
0             0           1  Chocolate Sandwich Cookies        61   
1             0           1  Chocolate Sandwich Cookies        61   
2             0           1  Chocolate Sandwich Cookies        61   
3             0           1  Chocolate Sandwich Cookies        61   
4             0           1  Chocolate Sandwich Cookies        61   

   department_id  prices  Unnamed: 0_y  order_id  user_id  order_number  ...  \
0             19     5.8          1987   3139998      138            28  ...   
1             19     5.8          1989   1977647      138            30  ...   
2             19     5.8         11433    389851      709             2  ...   
3             19     5.8         12198    652770      764             1  ...   
4             19     5.8         12200   1813452      764             3  ...   

   reordered  _merge    price_range_loc   busiest_day  busiest_days  \
0          0    both  Mid Range product  Regular_busy  Regular Busy   
1          1    both  Mid Range product  Regular_busy  Regular Busy   
2          0    both  Mid Range product   Busiest day  Busiest days   
3          0    both  Mid Range product  Regular_busy  Slowest days   
4          1    both  Mid Range product    Least busy  Slowest days   

  busiest_period_of_day max_order      loyalty_flag avg_price spending_flag  
0           Most orders        32  Regular Customer  6.935811   Low spender  
1           Most orders        32  Regular Customer  6.935811   Low spender  
2        Average orders         5      New customer  7.930208   Low spender  
3           Most orders         3      New customer  4.972414   Low spender  
4           Most orders         3      New customer  4.972414   Low spender  

[5 rows x 24 columns]

In [15]:
ords_prods_merge['spending_flag'].value_counts(dropna = False)

Low spender     31770746
High spender      634113
Name: spending_flag, dtype: int64

### 7)

In [16]:
# Create flag column of customers order frequency
ords_prods_merge['median_prior_order'] = ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform(np.median)

In [17]:
ords_prods_merge.loc[ords_prods_merge['median_prior_order'] > 20, 'frequency_flag'] = 'Non-frequent customer'
ords_prods_merge.loc[(ords_prods_merge['median_prior_order'] <= 20) & (ords_prods_merge['median_prior_order'] > 10), 'frequency_flag'] = 'Regular customer'
ords_prods_merge.loc[ords_prods_merge['median_prior_order'] <= 10, 'frequency_flag'] = 'Frequent Customer'

In [18]:
# Check the frequency flag
ords_prods_merge.head()

Unnamed: 0_x  product_id                product_name  aisle_id  \
0             0           1  Chocolate Sandwich Cookies        61   
1             0           1  Chocolate Sandwich Cookies        61   
2             0           1  Chocolate Sandwich Cookies        61   
3             0           1  Chocolate Sandwich Cookies        61   
4             0           1  Chocolate Sandwich Cookies        61   

   department_id  prices  Unnamed: 0_y  order_id  user_id  order_number  ...  \
0             19     5.8          1987   3139998      138            28  ...   
1             19     5.8          1989   1977647      138            30  ...   
2             19     5.8         11433    389851      709             2  ...   
3             19     5.8         12198    652770      764             1  ...   
4             19     5.8         12200   1813452      764             3  ...   

     price_range_loc   busiest_day  busiest_days  busiest_period_of_day  \
0  Mid Range product  Regular_busy  Regular Busy            Most orders   
1  Mid Range product  Regular_busy  Regular Busy            Most orders   
2  Mid Range product   Busiest day  Busiest days         Average orders   
3  Mid Range product  Regular_busy  Slowest days            Most orders   
4  Mid Range product    Least busy  Slowest days            Most orders   

   max_order      loyalty_flag avg_price spending_flag median_prior_order  \
0         32  Regular Customer  6.935811   Low spender                8.0   
1         32  Regular Customer  6.935811   Low spender                8.0   
2          5      New customer  7.930208   Low spender                7.0   
3          3      New customer  4.972414   Low spender                9.0   
4          3      New customer  4.972414   Low spender                9.0   

      frequency_flag  
0  Frequent Customer  
1  Frequent Customer  
2  Frequent Customer  
3  Frequent Customer  
4  Frequent Customer  

[5 rows x 26 columns]

In [20]:
ords_prods_merge['frequency_flag'].value_counts(dropna = False)

Frequent Customer        22790523
Regular customer          6927608
Non-frequent customer     2686728
Name: frequency_flag, dtype: int64

In [ ]:
# Export data
ords_prods_merge.to_pickle(os.path.join(path,'02 Data','Prepared Data','orders_products_merged.pkl'))